src/keys/slip10.rs

```diff
@@ -399,88 +433,89 @@ impl<K: hazmat::Derivable> TryFrom<&[u8; 65]> for Slip10<K> {
     }
 }
 
-#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
-pub struct Segment(pub u32);
-
-impl Segment {
-    pub fn is_hardened(&self) -> bool {
-        self.0 & Self::HARDEN_MASK != 0
+pub trait Segment: Copy + Into<u32> {
+    fn is_hardened(self) -> bool;
+    fn ser32(self) -> [u8; 4] {
+        self.into().to_be_bytes()
     }
-
-    pub fn bs(&self) -> [u8; 4] {
-        self.0.to_be_bytes() // ser32(i)
-    }
-
-    pub const HARDEN_MASK: u32 = 1 << 31;
+    fn harden(self) -> Hardened;
+    fn unharden(self) -> NonHardened;
 }
 
-impl From<u32> for Segment {
-    fn from(i: u32) -> Self {
-        Self(i)
-    }
-}
+const HARDEN_MASK: u32 = 1 << 31;
 
-impl From<Segment> for u32 {
-    fn from(s: Segment) -> Self {
-        s.0
+impl Segment for u32 {
+    fn is_hardened(self) -> bool {
+        self & HARDEN_MASK != 0
     }
-}
-
-#[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
-pub struct Chain(Vec<Segment>);
-
-impl Chain {
-    pub fn empty() -> Self {
-        Self(Vec::new())
+    fn harden(self) -> Hardened {
+        Hardened(self | HARDEN_MASK)
     }
-
-    pub fn from_segments<'a, I: IntoIterator<Item = &'a Segment>>(is: I) -> Self {
-        Self(is.into_iter().cloned().collect())
+    fn unharden(self) -> NonHardened {
+        NonHardened(self & !HARDEN_MASK)
     }
+}
 
-    pub fn from_u32<I: IntoIterator<Item = u32>>(is: I) -> Self {
-        Self(is.into_iter().map(Segment).collect())
-    }
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
+pub struct Hardened(pub u32);
 
-    pub fn from_u32_hardened<I: IntoIterator<Item = u32>>(is: I) -> Self {
-        Self::from_u32(is.into_iter().map(|i| Segment::HARDEN_MASK | i))
+impl From<Hardened> for u32 {
+    fn from(segment: Hardened) -> u32 {
+        segment.0
     }
+}
 
-    pub fn join<O: AsRef<Chain>>(&self, o: O) -> Self {
-        let mut ss = self.0.clone();
-        ss.extend_from_slice(&o.as_ref().0);
-        Self(ss)
+impl TryFrom<u32> for Hardened {
+    type Error = ();
+    fn try_from(segment: u32) -> Result<Self, ()> {
+        if segment.is_hardened() {
+            Ok(Hardened(segment))
+        } else {
+            Err(())
```

Comment:
Should return an actual Error and not just `()`?

Reply:
Probably should use crate::Error for that matter.

Reply:
What would be the best way to define an error?

Reply:
Added `SegmentHardeningError`. Added `Slip10Error` pattern to `crypto::Error` enum.